dbms/src/AggregateFunctions/AggregateFunctionGroupArray.h

```diff
@@ -43,51 +71,162 @@ struct GroupArrayNumericData
+    {
+        /// With a large number of values, we will generate random numbers several times slower.
+        if (lim <= static_cast<UInt64>(rng.max()))
+            return static_cast<UInt32>(rng()) % static_cast<UInt32>(lim);
+        else
+            return (static_cast<UInt64>(rng()) * (static_cast<UInt64>(rng.max()) + 1ULL) + static_cast<UInt64>(rng())) % lim;
+    }
+
+    void randomShuffle()
+    {
+        for (size_t i = 1; i < value.size(); ++i)
+        {
+            size_t j = genRandom(i + 1);
+            std::swap(value[i], value[j]);
+        }
+    }
+};
 
-template <typename T, typename Tlimit_num_elems>
+template <typename T, typename Trait>
 class GroupArrayNumericImpl final
-    : public IAggregateFunctionDataHelper<GroupArrayNumericData<T>, GroupArrayNumericImpl<T, Tlimit_num_elems>>
+    : public IAggregateFunctionDataHelper<GroupArrayNumericData<T, Trait::sampler != Sampler::NONE>, GroupArrayNumericImpl<T, Trait>>
 {
-    static constexpr bool limit_num_elems = Tlimit_num_elems::value;
+    using Data = GroupArrayNumericData<T, Trait::sampler != Sampler::NONE>;
+    static constexpr bool limit_num_elems = Trait::has_limit;
     DataTypePtr & data_type;
     UInt64 max_elems;
+    UInt64 seed;
 
 public:
-    explicit GroupArrayNumericImpl(const DataTypePtr & data_type_, UInt64 max_elems_ = std::numeric_limits<UInt64>::max())
-        : IAggregateFunctionDataHelper<GroupArrayNumericData<T>, GroupArrayNumericImpl<T, Tlimit_num_elems>>({data_type_}, {})
-        , data_type(this->argument_types[0]), max_elems(max_elems_) {}
+    explicit GroupArrayNumericImpl(const DataTypePtr & data_type_, UInt64 max_elems_ = std::numeric_limits<UInt64>::max(), UInt64 seed_ = 123456)
+        : IAggregateFunctionDataHelper<GroupArrayNumericData<T, Trait::sampler != Sampler::NONE>, GroupArrayNumericImpl<T, Trait>>(
+            {data_type_}, {})
+        , data_type(this->argument_types[0])
+        , max_elems(max_elems_)
+        , seed(seed_)
+    {
+    }
 
-    String getName() const override { return "groupArray"; }
+    String getName() const override { return getNameByTrait<Trait>(); }
 
     DataTypePtr getReturnType() const override
     {
         return std::make_shared<DataTypeArray>(data_type);
     }
 
+    void insert(Data & a, const T & v, Arena * arena) const
+    {
+        ++a.total_values;
+        if (a.value.size() < max_elems)
+            a.value.push_back(v, arena);
+        else
+        {
+            UInt64 rnd = a.genRandom(a.total_values);
+            if (rnd < max_elems)
+                a.value[rnd] = v;
+        }
+    }
+
     void add(AggregateDataPtr place, const IColumn ** columns, size_t row_num, Arena * arena) const override
     {
-        if (limit_num_elems && this->data(place).value.size() >= max_elems)
-            return;
+        if constexpr (Trait::sampler == Sampler::NONE)
+        {
+            if (limit_num_elems && this->data(place).value.size() >= max_elems)
```

Comment:
I also have a thought about "groupArrayLast" :)

Reply:
Could you elaborate? what elements should be victimized in such case

Reply:
groupArrayLast is intended to select "last" elements. It can be implemented as a cycle buffer.

Reply:
I found it useful for one of my use cases, so here it is - #44521